lib/src/test/java/io/ably/lib/test/realtime/RealtimePresenceTest.java

```diff
@@ -1649,6 +1649,125 @@ public void onPresenceMessage(PresenceMessage message) {
         }
     }
 
+    private final class CountingCompletionListener implements CompletionListener
+    {
+        public int successfulListeners = 0;
+        public int failedListeners = 0;
+
+        @Override
+        public void onSuccess() {
+            synchronized(this) {
+                successfulListeners++;
+                notifyAll();
+            }
+        }
+
+        @Override
+        public void onError(ErrorInfo reason) {
+            synchronized(this) {
+                failedListeners++;
+                notifyAll();
+            }
+        }
+    }
+
+    /**
+     * <p>
+     * Validates a client sending multiple presence updates when the channel is in the attaching
+     * state will have all messages sent once the channel attaches, and all listeners will be called.
+     * </p>
+     *
+     * @throws AblyException
+     */
+    @Test
+    public void realtime_presence_update_multiple_queued_messages() throws AblyException {
+        /* Ably instance that will emit presence events */
+        AblyRealtime ably1 = null;
+        /* Ably instance that will receive presence events */
+        AblyRealtime ably2 = null;
+
+        String channelName = "test.presence.subscribe.update_multiple_queued_messages" + System.currentTimeMillis();
+        EnumSet<PresenceMessage.Action> actions = EnumSet.of(Action.update, Action.enter);
+
+        try {
+            ClientOptions option1 = createOptions(testVars.keys[0].keyStr);
+            option1.clientId = "emitter client";
+            ClientOptions option2 = createOptions(testVars.keys[0].keyStr);
+            option2.clientId = "receiver client";
+
+            ably1 = new AblyRealtime(option1);
+            ably2 = new AblyRealtime(option2);
+
+            Channel channel1 = ably1.channels.get(channelName);
+
+            Channel channel2 = ably2.channels.get(channelName);
+            channel2.attach();
+            (new ChannelWaiter(channel2)).waitFor(ChannelState.attached);
+
+            CountingCompletionListener messageCompletionListener = new CountingCompletionListener();
+
+            final ArrayList<PresenceMessage> receivedMessageStack = new ArrayList<>();
+            channel2.presence.subscribe(actions, new Presence.PresenceListener() {
+                @Override
+                public void onPresenceMessage(PresenceMessage message) {
+                    synchronized (receivedMessageStack) {
+                        receivedMessageStack.add(message);
+                        receivedMessageStack.notify();
+                    }
+                }
+            });
+
+            /* Start emitting channel with ably client 1 (emitter) */
+            channel1.presence.enter("Hello, #2!", messageCompletionListener);
```

Comment:
Nothing is happening here to ensure that the enter/update calls all happen before the channel is attached. Do you have any thoughts on how to do that?

Reply:
It's quite a hard one to synchronize, but I believe it can be done by the following:

- Wrap the calls to presence.enter and presence.update in the test with a synchronize(channel1) call.
- setState (called by setAttached) also synchronizes on the channel, so we should prevent the channel from entering the attached state before we've had chance to sort out all the messages.